drivers/watchdog/iwdg_stm32.c

```diff
@@ -74,43 +80,48 @@ static const struct wdt_driver_api iwdg_stm32_api = {
 	.reload = iwdg_stm32_reload,
 };
 
-static inline int __iwdg_stm32_prescaler(int setting)
+static int iwdg_stm32_init(struct device *dev)
 {
-	int v;
-	int i = 0;
+	IWDG_TypeDef *iwdg = IWDG_STM32_STRUCT(dev);
+	u32_t tickstart;
+	u32_t prescaler = LL_IWDG_PRESCALER(CONFIG_IWDG_STM32_PRESCALER);
 
-	/* prescaler range 4 - 256 */
-	for (v = 4; v < 256; v *= 2, i++) {
-		if (v == setting)
-			return i;
-	}
-	return i;
-}
+	assert_param(LL_IS_IWDG_PRESCALER(prescaler));
+	assert_param(LL_IS_IWDG_RELOAD(CONFIG_IWDG_STM32_RELOAD_COUNTER));
 
-static int iwdg_stm32_init(struct device *dev)
-{
-	volatile struct iwdg_stm32 *iwdg = AS_IWDG(IWDG_BASE);
+	LL_IWDG_Enable(iwdg);
+	LL_IWDG_EnableWriteAccess(iwdg);
 
-	/* clock setup is not required, once the watchdog is enabled
-	 * LSI oscillator will be forced on and fed to IWD after
-	 * stabilization period
-	 */
+	LL_IWDG_SetPrescaler(iwdg, prescaler);
+	LL_IWDG_SetReloadCounter(iwdg, CONFIG_IWDG_STM32_RELOAD_COUNTER);
+	LL_IWDG_SetWindow(iwdg, 0x0FFF);
 
-	/* unlock access to configuration registers */
-	iwdg->kr.bit.key = STM32_IWDG_KR_UNLOCK;
+	tickstart = k_uptime_get_32();
 
-	iwdg->pr.bit.pr =
-		__iwdg_stm32_prescaler(CONFIG_IWDG_STM32_PRESCALER);
-	iwdg->rlr.bit.rl = CONFIG_IWDG_STM32_RELOAD_COUNTER;
+	while (LL_IWDG_IsReady(iwdg) == 0) {
+		if ((k_uptime_get_32() - tickstart) > IWDG_DEFAULT_TIMEOUT) {
+			return -ENODEV;
+		}
+	}
+
+	LL_IWDG_ReloadCounter(iwdg);
 
-#ifdef CONFIG_IWDG_STM32_START_AT_BOOT
-	iwdg_stm32_enable(dev);
-#endif
+	/*
+	 * As LSI is turned on automatically,
+	 * IWDG is enabled by default.
```

Comment:
The ST production value for the option bytes where WDG_SW bit is present is 0x00FF55AA, namely the Software watchdog mode is enabled by default.
If the Independent watchdog (IWDG) is started by either hardware option or software access, the LSI oscillator is forced ON and cannot be disabled. So there is no need to take care about LSI first.
I will provide more precise comment.